source/code/plugins/in_npmd.rb

```diff
@@ -0,0 +1,676 @@
+                        _agent["IPs"] << _tempIp
+                    end
+                    _a << _agent
+                end
+                _a
+            rescue JSON::ParserError => e
+                Logger::logError "Error in Json Parse in agent data: #{e}", Logger::resc
+                nil
+            end
+        end
+
+        def self.getRuleHashFromJson(text)
+            begin
+                _h = JSON.parse(text)
+                _a = Array.new
+                _h.each do |key, value|
+                    next if value.nil? or
+                        !(value["ActOn"].is_a?Array) or
+                        !(value["Exceptions"].is_a?Array)
+                    _rule = Hash.new
+                    _rule["Name"] = key
+                    _rule["LossThreshold"] = value["Threshold"]["Loss"]
+                    _rule["LatencyThreshold"] = value["Threshold"]["Latency"]
+                    _rule["Protocol"] = value["Protocol"] unless value["Protocol"].nil?
+                    _rule["Rules"] = value["ActOn"]
+                    _rule["Exceptions"] = value["Exceptions"]
+                    _a << _rule
+                end
+                _a
+            rescue JSON::ParserError => e
+                Logger::logError "Error in Json Parse in rule data: #{e}", Logger::resc
+                nil
+            end
+        end
+
+    end
+
+    # Only function needed to be called from this module
+    def self.GetAgentConfigFromUIConfig(uiXml)
+        begin
+            _uiHash = UIConfigParser.parse(uiXml)
+            _agentXml = AgentConfigCreator.createXmlFromUIConfigHash(_uiHash)
+            nil
+        rescue Exception => e
+            Logger::logError "Got exception as: #{e}"
+            raise e
+        end
+    end
+end
+
+# The npmd input plugin to fluentd
+module Fluent
+    class NPM < Input
+        Fluent::Plugin.register_input('npmd', self)
+
+        unless method_defined?(:router)
+            define_method("router") {Fluent::Engine}
+        end
+
+        def initialize
+            super
+            require 'socket'
+            require 'fileutils'
+            require 'json'
+
+            require_relative 'oms_common'
+            require_relative 'oms_configuration'
+        end
+
+        def configure(conf)
+            super
+        end
+
+        Cmd_start         = "StartNPMD"
+        Cmd_stop          = "StopNPMD"
+        Cmd_config        = "ConfigNPMD"
+        Cmd_restart       = "RestartNPMD"
+        Cmd_log           = "ErrorLog"
+
+        Npmd_conn_confirm = "NPMDAgent Connected!"
+
+        Loc_unix_endpoint = "/var/opt/microsoft/omsagent/run/npmdagent.sock"
+        Loc_config_file   = "/etc/opt/microsoft/omsagent/conf/npmdAgentConfig.conf"
+        Loc_npmd_binary   = "/opt/microsoft/omsagent/Daemon"
+
+        Stop_signal       = "SIGKILL"
+        Tag               = "npmd.oms.blob"
+
+        Retry_delay_long  = 900 # 15 minutes
```

Comment:
please append min or sec keyword to the parameter names so that it is clear which one is which